### ISSUE TITLE
multiple termQuery not working

### ISSUE THREAD

**Issue by dijadev** [ongr-io/ElasticsearchDSL#179]
Hi !
I am trying to execute this query but it seems that it's not supported to add multiple termQuery.
 
```
<?php
       $search = new Search();
        $termQueryForTag1 = new TermQuery("tag", "wow");
        $termQueryForTag2 = new TermQuery("tag", "elasticsearch");
        $termQueryForTag3 = new TermQuery("tag", "dsl");

        $search->addQuery($termQueryForTag1);
        $search->addQuery($termQueryForTag2);
        $search->addQuery($termQueryForTag3, BoolQuery::MUST);

        $queryArray = $search->toArray();
        echo "<pre>";var_dump($queryArray);die(); 
```

Just the laste query was token 

```
array(1) {
  ["query"]=>
  array(1) {
    ["term"]=>
    array(1) {
      ["tag"]=>
      string(3) "dsl"
    }
  }
} 
```
Is it a bug or I am doing it wrongly ?

PS: I am using ElasticSearch 2.3.3 and elasticsearch-dsl 2.2.

Thanks in advance,

**Comment by saimaz** [comment]
Hi @dijadev 

The query is correct. But in your case it should be formed as 3 bool queries in the must operator.

```json
{
 "query": {
 "bool":
  {
   "must":[
      {
         "term":{"tag":"wow"}
      },
      {
        "term":{"tag":"elasticsearch"}
      },
      { 
        "term":{"tag":"dsl"}
      }
    ]
  }
}
}
```

**Comment by dijadev** [comment]
Hi @saimaz  thank you for replying!
I tested by adding 3 bool queries but it does not work
Could you give the right query to get the array you showed ?

**Comment by saimaz** [comment]
You query is right. The output should be like I posted. 

```php
        $search = new Search();
        $termQueryForTag1 = new TermQuery("tag", "wow");
        $termQueryForTag2 = new TermQuery("tag", "elasticsearch");
        $termQueryForTag3 = new TermQuery("tag", "dsl");

        $search->addQuery($termQueryForTag1);
        $search->addQuery($termQueryForTag2);
        $search->addQuery($termQueryForTag3);

        $queryArray = $search->toArray();
        echo "<pre>";json_encode($queryArray);die(); 
```

Generates:
```json
{
 "query": {
 "bool":
  {
   "must":[
      {
         "term":{"tag":"wow"}
      },
      {
        "term":{"tag":"elasticsearch"}
      },
      { 
        "term":{"tag":"dsl"}
      }
    ]
  }
}
}
```

If not, sent me more info about your env. The exact ElasticsearchDSL version 2.2.x (what is x number), your PHP version.

**Comment by dijadev** [comment]
I checked the code in the bundle and i found a key optional arg so i added for each query a different key as bellow and it works !

```
 $search = new Search();
        $termQueryForTag1 = new TermQuery("tag", "wow",1);
        $termQueryForTag2 = new TermQuery("tag", "elasticsearch",2);
        $termQueryForTag3 = new TermQuery("tag", "dsl",3);

        $search->addQuery($termQueryForTag1);
        $search->addQuery($termQueryForTag2);
        $search->addQuery($termQueryForTag3);

        $queryArray = $search->toArray();
        echo "<pre>";json_encode($queryArray);die(); 
```
This must be fixed or montioned in the doc.

Thanks,

**Comment by saimaz** [comment]
It's still very interesting why key is not generated and when you add the second query it drops first. I still cannot reproduce your issue :(. 

The key is for case when you want to add some identifier where you later can grab that specific query from `BoolQuery` container. It should not be necessary for adding multiple `TermQuery`.

**Comment by dijadev** [comment]
Ok I'll debug this as soon as i can and will keep you in touch :)

**Comment by puskic** [comment]
Probably the problem is uniqid() in BoolQuery where he is assigning the key if not exists. I have the same problem on my local machine(xampp, php 7, windows 7) where microtime is just not refreshing quick enough. I solve it with  hardcoded random generator


    public function add(BuilderInterface $query, $type = self::MUST, $key = null)
    {
        if (!in_array($type, [self::MUST, self::MUST_NOT, self::SHOULD, self::FILTER])) {
            throw new \UnexpectedValueException(sprintf('The bool operator %s is not supported', $type));
        }

        if (!$key) {
            $key = str_random(60);          //uniqid();
        }

        $this->container[$type][$key] = $query;

        return $key;
    }
 
    public static function random($length = 16)
    {
        $string = '';

        while (($len = strlen($string)) < $length) {
            $size = $length - $len;

            $bytes = static::randomBytes($size);

            $string .= substr(str_replace(['/', '+', '='], '', base64_encode($bytes)), 0, $size);
        }

        return $string;
    }

**Comment by saimaz** [comment]
Thank you @puskic. The `uniqid()` is definitely a problem. Your provided solution would work, but it seems a bit overhead to calculate just a unique string. I believe there should be more optimized way to do this.

**Comment by puskic** [comment]
I agree with you. I already have it, way not to use ;). 
Probably 

bin2hex(random_bytes(30))  

will do the work.
Strange result will came out if $key overlaps, it should be unique enough...

**Comment by dijadev** [comment]
Hi !

I confirm what @puskic said; this problem occurs just in windows env, and disappears when I deploy to a linux machine..
```
    public function add(BuilderInterface $query, $type = self::MUST, $key = null)
    {
        if (!in_array($type, [self::MUST, self::MUST_NOT, self::SHOULD, self::FILTER])) {
            throw new \UnexpectedValueException(sprintf('The bool operator %s is not supported', $type));
        }

        if (!$key) {
            $key = uniqid();
            echo $type.'=>'.$key.'<br>';
        }

        $this->container[$type][$key] = $query;

        return $key;
    }
```
It generates the same $key for the TermQuerys I added to BoolQuery 
```
must=>5860f2aa7f977
must=>5860f2aa9fd1f
must=>5860f2aabc62e
must=>5860f2aabc62e
must=>5860f2aabcdfe
```

**Comment by saimaz** [comment]
Fixed in v2.2.1 release
